push-down-test/src/main.go

```diff
@@ -0,0 +1,260 @@
+	ok := diffRunResult(testCasePath, noPushDownLogChan, pushDownLogChan, pushDownWithBatchLogChan)
+	if !ok {
+		log.Printf("Test failed. Stop.")
+		os.Exit(2) // Diff fail results in exit code 2 to distinguish with panic
+	}
+}
+
+func runStatements(logChan chan *statementLog, connString string, statements []string) {
+	db := mustDBOpen(connString)
+	mustDBExec(db, "use `"+*dbName+"`;")
+	for i, stmt := range statements {
+		runSingleStatement(stmt, i, db, logChan)
+	}
+	mustDBClose(db)
+	close(logChan)
+}
+
+func runSingleStatement(stmt string, stmtIndex int, db *sql.DB, logChan chan *statementLog) bool {
+	hasError := false
+	logBuf := new(bytes.Buffer)
+	rows, err := db.Query(stmt)
+	if err != nil {
+		hasError = true
+		logBuf.WriteString(string(err.Error()))
+		logBuf.WriteString("\n")
+	} else {
+		cols, err := rows.Columns()
+		expectNoErr(err)
+		if len(cols) > 0 {
+			byteRows, err := SqlRowsToByteRows(rows)
+			expectNoErr(err)
+			WriteQueryResult(byteRows, logBuf)
+		}
+		logBuf.WriteString("\n")
+		expectNoErr(rows.Close())
+		expectNoErr(rows.Err())
+	}
+	logChan <- &statementLog{
+		output:    logBuf,
+		stmt:      stmt,
+		stmtIndex: stmtIndex,
+		hasError:  hasError,
+	}
+	return !hasError
+}
+
+func diffRunResult(testCasePath string,
+	noPushDownLogChan chan *statementLog,
+	pushDownLogChan chan *statementLog,
+	pushDownWithBatchLogChan chan *statementLog) bool {
+
+	execOkStatements := 0
+	execFailStatements := 0
+	diffFailStatements := 0
+	successQueries := new(bytes.Buffer)
+
+	for {
+		noPushDownLog, ok1 := <-noPushDownLogChan
+		pushDownLog, ok2 := <-pushDownLogChan
+		pushDownWithBatchLog, ok3 := <-pushDownWithBatchLogChan
+
+		allEnd := !(ok1 || ok2 || ok3)
+		if allEnd {
+			break
+		}
+		if !ok1 {
+			log.Panicf("Internal error: NoPushDown channel drained\n")
+		}
+		if !ok2 {
+			log.Panicf("Internal error: PushDown channel drained\n")
+		}
+		if !ok3 {
+			log.Panicf("Internal error: PushDownWithBatch channel drained\n")
+		}
+		if noPushDownLog.stmt != pushDownLog.stmt ||
+			pushDownLog.stmt != pushDownWithBatchLog.stmt {
+			log.Panicln("Internal error: Pre-check failed, stmt should be identical",
+				noPushDownLog.stmt, pushDownLog.stmt, pushDownWithBatchLog.stmt)
+		}
+		if noPushDownLog.stmtIndex != pushDownLog.stmtIndex ||
+			pushDownLog.stmtIndex != pushDownWithBatchLog.stmtIndex {
+			log.Panicln("Internal error: Pre-check failed, stmtIndex should be identical",
+				noPushDownLog.stmtIndex, pushDownLog.stmtIndex, pushDownWithBatchLog.stmtIndex)
+		}
+
+		if noPushDownLog.hasError || pushDownLog.hasError || pushDownWithBatchLog.hasError {
+			execFailStatements++
+		} else {
+			execOkStatements++
+		}
+
+		if !bytes.Equal(noPushDownLog.output.Bytes(), pushDownLog.output.Bytes()) ||
+			!bytes.Equal(pushDownLog.output.Bytes(), pushDownWithBatchLog.output.Bytes()) {
+			diffFailStatements++
+			log.Printf("Test fail: Outputs are not matching.\n"+
```

Comment:
Add one `\n` at the end of this output. Or the output is not easy to read.

![image](https://user-images.githubusercontent.com/21231345/66982921-de5a0b00-f0e9-11e9-9f99-f0d37f1d1fe1.png)

Reply:
Also, add \n at the end of each log.